src/main/asciidoc/repository-projections.adoc

```diff
@@ -1,163 +1,276 @@
+  default String getFullName() {
+    return getFirstname.concat(" ").concat(getLastname());
+  }
+}
+----
+====
 
-Some applications require the full name of a person. Concatenating strings with `String.format("%s %s", person.getFirstName(), person.getLastName())` would be one possibility but this piece of code needs to be called in every place the full name is required. Virtual properties on projections leverage the need for repeating that code all over.
+This approach requires you to be able to implement logic purely based on the other accessor methods exposed on the projection interface.
+A second, more flexible option is to implement the custom logic in a Spring bean and then simply invoke that from the SpEL expression:
 
-[source,java]
+[[projections.interfaces.open.bean-reference]]
+.Sample Person object
+====
+[source, java]
 ----
-interface FullNameAndCountry {
+@Component
+class MyBean {
 
-  @Value("#{target.firstName} #{target.lastName}")
+  String getFullName(Person person) {
+    …
+  }
+}
+
+interface NamesOnly {
+
+  @Value("#{@myBean.getFullName(target)}")
   String getFullName();
+  …
+}
+----
+====
+
+Note, how the SpEL expression refers to `myBean` and invokes the `getFullName(…)` method forwarding the projection target as method parameter.
+Methods backed by SpEL expression evaluation can also use method parameters which can then be referred to from the expression.
+The method parameters are available via an `Object` array named `args`.
 
-  @Value("#{target.address.country}")
-  String getCountry();
+.Sample Person object
+====
+[source, java]
+----
+interface NamesOnly {
+
+  @Value("#{args[0] + ' ' + target.firstname + '!'}")
+  String getSalutation(String prefix);
 }
 ----
+====
 
-In fact, `@Value` gives full access to the target object and its nested properties. SpEL expressions are extremly powerful as the definition is always applied to the projection method. Let's take SpEL expressions in projections to the next level.
+Again, for more complex expressions rather use a Spring bean and let the expression just invoke a method as described in <<projections.interfaces.open.bean-reference, above>>.
 
+[[projections.dtos]]
+== Class-based projections (DTOs)
 
-Imagine you had the following domain model definition:
+Another way of defining
 
-[source,java]
+.A projecting DTO
+====
+[source, java]
 ----
-@Entity
-public class User {
+class NamesOnly {
 
-  @Id @GeneratedValue
-  private Long id;
-  private String name;
+  private final String firstname, lastname;
 
-  private String password;
-  …
+  NamesOnly(String firstname, String lastname) {
+    this.firstname = firstname;
+    this.lastname = lastname;
+  }
+
+  String getFirstname() {
+    return this.firstname;
+  }
+
+  String getLastname() {
+    return this.lastname;
+  }
+
+  // equals(…) and hashCode() implementations
+}
+----
+====
+
+[TIP]
+.Avoiding boilerplate code for projection DTOs
```

Comment:
Not sure if we should advertise Lombok in our documentation. There are other libraries out there doing similar things and SD doesn't really care, does it?